### ISSUE TITLE
Setting Emissary as the default Ingress controller ignores created Ingress objects.

### ISSUE THREAD

**Issue by james-masson** [emissary-ingress/emissary#5194]
**Describe the bug**

When installed via Helm with the flag `ingressClassResource.default = true` , emissary/ambassador correctly sets itself as default ingress.

```
$ kubectl get ingressclass ambassador -o yaml
apiVersion: networking.k8s.io/v1
kind: IngressClass
metadata:
  annotations:
    ingressclass.kubernetes.io/is-default-class: "true"
    meta.helm.sh/release-name: emissary-ingress
    meta.helm.sh/release-namespace: emissary-ingress
  creationTimestamp: "2023-07-24T09:42:10Z"
  generation: 1
  labels:
    app.kubernetes.io/managed-by: Helm
  name: ambassador
  resourceVersion: "60464"
  uid: 790bb84c-7138-4284-a0b6-8d1b3e2852d1
spec:
  controller: getambassador.io/ingress-controller

```

But any created Ingress definitions are ignored by Emissary, unless the Ingress has `ingressClassName: ambassador` set.

This negates the purpose of a default Ingress controller.

**To Reproduce**

1. Install emissary with `ingressClassResource.default = true` 
2. Create an ingress definition that doesn't use a specified ingress definition

eg.
```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  annotations:
    meta.helm.sh/release-name: admin-gw
    meta.helm.sh/release-namespace: local
  creationTimestamp: "2023-07-25T10:08:55Z"
  generation: 7
  labels:
    app.kubernetes.io/instance: admin-gw
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: app
    app.kubernetes.io/version: 4.16.0
    helm.sh/chart: app-0.0.1
  name: admin-gw
  namespace: local
  resourceVersion: "61581"
  uid: b2baa14a-cb94-4c9c-b0ef-ea20f5b4a48c
spec:
  rules:
  - host: admin-gw.app.local
    http:
      paths:
      - backend:
          service:
            name: admin-gw
            port:
              number: 8929
        path: /
        pathType: Prefix
status:
  loadBalancer: {}
```

3. Ingress definitions are ignored, and fail to show up on Ambassador Admin page.

**Expected behavior**

Setting Emissary/Ambassador as default Ingress should mean that all Ingress definitions can use it without specifically selecting it.

**Versions:**
 - Ambassador: 3.7.1
 - Kubernetes environment: Minkube v1.30.1 - K8s version v1.25.6
 - Version: emissary-ingress-8.7.1

**Comment by james-masson** [comment]
This seems to be the problem.

https://github.com/emissary-ingress/emissary/blob/6ab7c1a40d6c99321707b9140da3ed53643d052e/python/ambassador/fetch/ingress.py#L130

Doesn't seem to take into account the case that Ambassador/Emissary is the default.

**Comment by cindymullins-dw** [comment]
I see [in the docs](https://www.getambassador.io/docs/emissary/latest/topics/running/ingress-controller) these two points:
- You must create your Ingress resource with the correct ingress.class. (I'm not actually sure what 'correct' means here.)
- Emissary-ingress will automatically read Ingress resources with the annotation kubernetes.io/ingress.class: ambassador.

It also notes that "The Ingress specification is very basic and does not support many of the features of Emissary-ingress, so you'll be using both the Ingress resource and Emissary-ingress's Mapping resource to manage your Kubernetes services.

Would the goal in this case be that you want to use Emissary as the default Ingress resource without having to give it the annotation `kubernetes.io/ingress.class: ambassador`?

**Comment by james-masson** [comment]
> You must create your Ingress resource with the correct ingress.class. (I'm not actually sure what 'correct' means here.)

This negates the point of setting the Emissary-ingress to be the default class. Either it's default and you don't have to specify the ingress class in the Ingress definition, or it's not the default class and you *do* have to tell Emissary "this one's for you". AFAIK this is the *only* use of setting an Ingress Controller to be default.

> Emissary-ingress will automatically read Ingress resources with the annotation kubernetes.io/ingress.class: ambassador.

This annotation is deprecated - see https://kubernetes.io/docs/concepts/services-networking/ingress/#deprecated-annotation

`ingressClassName:` is the replacement 

> It also notes that "The Ingress specification is very basic and does not support many of the features of Emissary-ingress, so you'll be using both the Ingress resource and Emissary-ingress's Mapping resource to manage your Kubernetes services.

Sort of true - you do need to specify a host - we use a wildcard. But the point of the Emissary-Ingress-Controller is to dynamically construct the mapping from the Ingress definition, which it does do successfully.

> Would the goal in this case be that you want to use Emissary as the default Ingress resource without having to give it the annotation kubernetes.io/ingress.class: ambassador?

Yes ( substitute `ingressClassName` for annotation ): This is the *only* point AFAIK of the flag `ingressClassResource.default = true`

**Comment by james-masson** [comment]
here's the K8s docs about what setting your ingress controller to default means: https://kubernetes.io/docs/concepts/services-networking/ingress/#default-ingress-class

> You can mark a particular IngressClass as default for your cluster. Setting the ingressclass.kubernetes.io/is-default-class annotation to true on an IngressClass resource will ensure that new Ingresses without an ingressClassName field specified will be assigned this default IngressClass.

**Comment by cindymullins-dw** [comment]
I will mark as a bug since subsequent ingress definitions should use the default once set.  Thanks for the link.  Just curious, rather than implementing Emissary as an "Ingress" resource, have you tried configuring Emissary [using CRDs](https://www.getambassador.io/docs/emissary/latest/topics/running/ingress-controller#when-to-use-an-ingress-instead-of-annotations-or-crds) as recommended?

**Comment by james-masson** [comment]
> Just curious, rather than implementing Emissary as an "Ingress" resource, have you tried configuring Emissary [using CRDs](https://www.getambassador.io/docs/emissary/latest/topics/running/ingress-controller#when-to-use-an-ingress-instead-of-annotations-or-crds) which is generally preferable if thats an option for you.

Yes, we tend to use Mappings in our bespoke applications where we want greater control.

However, we use large numbers of internet/upstream Helm Charts that have no support for Emissary Mappings, and using Ingress with them is much easier to maintain.

**Comment by eevdev** [comment]
@james-masson when I create an Ingress without ingressClass and there's a default IngressClass in the cluster, k8s will automatically add the default class to the Ingress, at least that's what's happening for me. I was under the impression that's handled by k8s somehow and not the Ingress controllers (Emissary/Nginx etc), but I might be wrong. Have you tried it on something other than Minikube?

Could you confirm that you only have one default IngressClass? `kubectl get -oyaml ingressclass | grep 'is-default-class'` should return maximum 1 line.

On a related note, when using AMBASSADOR_ID the `getambassador.io/ambassador-id` annotation is currently needed on both Ingress and IngressClass. It would be great if it was sufficient to set that on either and not necessarily both, for the same reason as @james-masson mentioned in his previous comment (upstream Ingresses can't always be easily configured).

See also https://github.com/emissary-ingress/emissary/issues/4691#issuecomment-1678520598
